doc/values.md

```diff
@@ -0,0 +1,210 @@
+
+![object-model](./assets/object-model.png)
+
+That heap data is managed by the JS engine, which means that the data is MOVED by the engine without notice. It is also actively cleaned up by the almighty garbage collector. This is a process that can be fraught for native code because it runs outside of the JS context.
+
+Because data is heap allocated, values provide two functions.  They facilitate lookup and garbage collection and there are two different systems for accomplishing that.  `v8::Context` is what [creates](./environment.md#) lexical context.  `internal::v8::Object` [doclink](https://github.com/nodejs/node/blob/4166d40d0873b6d8a0c7291872c8d20dc680b1d7/deps/v8/src/objects/objects.h#L299) and handles to those objects allow the runtime to function with data that moves.
+
+You can find the full inheritance hierarchy [here](./reference.md#object-inheritance-hierarchy), but pay attention to the following section as it will help to understand how `HandleScope` and `napi_value`  relate.
+
+```c++
+// - Object
+//   - Smi          (immediate small integer)
+//   - TaggedIndex  (properly sign-extended immediate small integer)
+//   - HeapObject   (superclass for everything allocated in the heap)
+//     - JSReceiver  (suitable for property access) ie. Object/Proxy
+//     - FixedArrayBase  (superclass for BufferLike and hash-table based data)
+//     - PrimitiveHeapObject  (primitives String/Symbol/Number)
+```
+
+The `Isolate` is responsible for handling all data related to JS execution and values are only the slice that addon developers interact with.  Memory is allocated in the heap to manage all things JS and native code handles  are references, to pointers, to said data.
+
+When working with `napi_value`, and by extension `Napi::Value`, it is important to pass by value.  While it is possible to also pass by reference, which is glorified pass by value, one should never pass by pointer. That insinuates some form of heap allocation for the `napi_value` and the value is not properly counted in the references because of the extra level of indirection.
+
+It is important to note that `v8::Object` and a JavaScript `Object` are not the same. In a JS script, `new Object` creates an instance of `JSObject` under-the-hood and that is a subclass of `JSReceiver`.
+
+The isolate creates objects during runtime and they will be stored in the faster to access space. The runtime will retain a `Handle` to the data so values can be looked up. The `Handle` will be added to, and managed by, the nearest `HandleScope` so that the GC can cleanup once the `HandleScope` is destroyed (variable goes out of scope lexically).
```

Comment:
```suggestion
The isolate creates objects during runtime and they will be stored in the faster-to-access space. The runtime will retain a `Handle` to the data so values can be looked up. The `Handle` will be added to, and managed by, the nearest `HandleScope` so that the GC can cleanup once the `HandleScope` is destroyed (variable goes out of scope lexically).
```

Reply:
Updated.